wpilib/tests/test_relay.py

```diff
@@ -155,13 +170,60 @@ def test_relay_get(dir, value, fwd, rev, wpilib, relay_data):
     #hal.getRelayForward.assert_called_once_with(relay._port)
     #hal.getRelayReverse.assert_called_once_with(relay._port)
 
+
 def test_relay_get_freed(relay, hal):
     relay.free()
     #hal.reset_mock()
     with pytest.raises(ValueError):
         relay.get()
     #assert not hal.setRelayForward.called and not hal.setRelayReverse.called
 
+
+def test_relay_getChannel(relay):
+    assert relay.getChannel() == 2
+
+
+def test_relay_expiration(relay):
+    relay.setExpiration(111)
+    assert relay.getExpiration() == 111
+
+
+def test_relay_isAlive1(relay):
+    relay.setSafetyEnabled(False)
+    assert relay.isAlive()
+    relay.setSafetyEnabled(True)
+    assert not relay.isAlive()
+
+
+def test_relay_isAlive2(wpilib):
+    with patch('hal_impl.functions.hooks', new=SimHooks()) as hooks:
+        hal_impl.functions.reset_hal()
+
+        relay = wpilib.Relay(2)
+
+        relay.setSafetyEnabled(True)
+        hooks.time = 1.0
+        assert not relay.isAlive()
+        relay.feed()
+        assert relay.isAlive()
+    
+
+def test_relay_stopMotor(relay, wpilib):
+    relay.set = MagicMock()
```

Comment:
In general, my preference would be that we avoid MagicMock-based tests and use the hal_data interfaces instead. I realize there are a bunch of them already, but a lot of those were created before the simulated hal existed and we've never gotten around to changing them. The advantage (in my opinion) of using the hal_data stuff is that it tests our hal boundaries too in addition to testing the wpilib stuff (and there is some typechecking done there as well). Where it's difficult to use, we should improve (or make helpers) so that it's easier for others to use too.

However, if you don't want to change the ones you've already done, that's fine, I will still merge these.

Reply:
I agree, but in this particular case, set's effects are a bit opaque, and test_relay_set covers it thoroughly

Reply:
Fair enough.